// type/src/index.js
var type = module.exports,

    toString = Object.prototype.toString,
    fnToString = Function.prototype.toString,

    HOST_CTOR = /^\[object .+?Constructor\]$/,

    IS_NATIVE = RegExp("^" +
        fnToString.call(toString)
        .replace(/[.*+?^${}()|[\]\/\\]/g, "\\$&")
        .replace(/toString|(function).*?(?=\\\()| for .+?(?=\\\])/g, "$1.*?") + "$"
    );


type.isError = function(obj) {

    return toString.call(obj) === "[object Error]";
};

type.isObject = function(obj) {
    var typeStr;

    if (obj === null || obj === undefined) return false;
    typeStr = typeof(obj);

    return (typeStr === "function" || typeStr === "object");
};

type.isHash = function(obj) {

    return toString.call(obj) === "[object Object]";
};

type.isArrayLike = function(obj) {

    return obj !== null && !type.isWindow(obj) && typeof(obj) === "object" && (obj.length === +obj.length);
};

type.isArray = Array.isArray || function isArray(obj) {

    return toString.call(obj) === "[object Array]";
};

type.isArguments = function(obj) {

    return toString.call(obj) === "[object Arguments]";
};

if (typeof(/./) !== "function") {
    type.isFunction = function(obj) {

        return typeof(obj) === "function";
    };
} else {
    type.isFunction = function(obj) {

        return toString.call(obj) === "[object Function]";
    };
}

type.isString = function(obj) {
    var type = typeof(obj);

    return type === "string" || (obj && type === "object" && toString.call(obj) === "[object String]") || false;
};

type.isNumber = function(obj) {
    var type = typeof(obj);

    return type === "number" || (obj && type === "object" && toString.call(obj) === "[object Number]") || false;
};

type.isFinite = Number.isFinite || function isFinite(obj) {
    return !(
        typeof(obj) !== "number" ||
        (obj !== obj || obj === Infinity || obj === -Infinity) ||
        false
    );
};

type.isNaN = Number.isNaN || function(obj) {

    return type.isNumber(obj) && (obj !== +obj);
};

type.isDecimal = function(obj) {

    return type.isNumber(obj) && obj % 1 !== 0;
};

type.isFloat = type.isDecimal;

type.isInteger = function(obj) {

    return type.isNumber(obj) && obj % 1 === 0;
};

type.isInt = type.isInteger;

type.isDate = function(obj) {

    return toString.call(obj) === "[object Date]";
};

type.isRegExp = function(obj) {

    return toString.call(obj) === "[object RegExp]";
};

type.isBoolean = function(obj) {

    return obj === true || obj === false || toString.call(obj) === "[object Boolean]";
};

type.isNull = function(obj) {

    return obj === null;
};

type.isUndefined = function(obj) {

    return obj === undefined;
};

type.isUndefinedOrNull = function(obj) {

    return obj === null || obj === undefined;
};

type.isDefined = function(obj) {

    return typeof(obj) !== "undefined";
};

type.isPrimitive = function(obj) {
    var typeStr = typeof(obj);
    return (obj === undefined || obj === null) || (typeStr !== "object" && typeStr !== "function");
};

type.isNative = function(obj) {
    var typeStr = typeof(obj);

    return (
        typeStr === "function" ?
        IS_NATIVE.test(fnToString.call(obj)) :
        (obj && typeStr === "object" && HOST_CTOR.test(toString.call(obj))) || false
    );
};

type.isNode = function(obj) {
    return obj != null && (
        typeof(Node) === "function" ? obj instanceof Node : (
            typeof(obj) === "object" &&
            typeof(obj.nodeType) === "number" &&
            typeof(obj.nodeName) === "string"
        )
    );
};

type.isTextNode = function(obj) {
    return type.isNode(obj) && obj.nodeType === 3;
};

type.isTextInputElement = (function() {
    var supportedInputTypes = {
        color: true,
        date: true,
        datetime: true,
        "datetime-local": true,
        email: true,
        month: true,
        number: true,
        password: true,
        range: true,
        search: true,
        tel: true,
        text: true,
        time: true,
        url: true,
        week: true
    };

    return function isTextInputElement(obj) {
        return obj != null && (
            (obj.nodeName === "INPUT" && supportedInputTypes[obj.type]) ||
            obj.nodeName === "TEXTAREA"
        );
    };
}());

type.isElement = function(obj) {
    return obj != null && (
        typeof(HTMLElement) === "function" ? obj instanceof HTMLElement : (
            typeof(obj) === "object" &&
            obj.nodeType === 1 &&
            typeof(obj.nodeName) === "string"
        )
    );
};

type.isWindow = function(obj) {

    return obj != null && obj.window === obj;
};

type.isDocument = function(obj) {

    return obj != null && (obj.constructor === Document || obj.constructor === HTMLDocument);
};

type.isEventTarget = function(obj) {
    return obj != null && (
        typeof(EventTarget) === "function" ? obj instanceof EventTarget : (
            type.isFunction(obj.addEventListener) ||
            type.isFunction(obj.attachEvent)
        )
    );
};
